app-backend/common/src/main/scala/geotrellis/spark/io/postgres/PostgresAttributeStore.scala

```diff
@@ -1,68 +1,124 @@
+  }
 
-  def write[T: JsonFormat](layerId: LayerId, attributeName: String, value: T): Unit =
+  def write[T: JsonFormat](layerId: LayerId, attributeName: String, value: T): Unit = {
+    logger.debug(s"write($layerId-$attributeName)")
     Await.result(LayerAttributes.insertLayerAttribute(
       LayerAttribute(
         layerName = layerId.name,
         zoom = layerId.zoom,
         name = attributeName,
         value = parse(value.toJson.toString).valueOr(throw _)
       )
-    ), Duration.Inf)
+    ), awaitTimeout)
+  }
+
+  def getHistogram[T: JsonFormat](layerId: LayerId): Future[Option[T]] = {
+    logger.debug(s"getHistogram($layerId)")
+    LayerAttributes.read(layerId.name, layerId.zoom, "histogram").map { opt =>
+      Option(opt
+        .map(_.value)
+        .getOrElse(throw new Exception(s"No attribute: $layerId, Histogram"))
+        .noSpaces
+        .parseJson
+        .convertTo[T])
+    }
+  }
 
-  def layerExists(layerId: LayerId): Boolean =
-    Await.result(LayerAttributes.layerExists(layerId.name, layerId.zoom), Duration.Inf)
+  def layerExists(layerId: LayerId): Boolean = {
+    logger.debug(s"layerExists($layerId)")
+    Await.result(LayerAttributes.layerExists(layerId.name, layerId.zoom), awaitTimeout)
+  }
 
-  def delete(layerId: LayerId): Unit =
-    Await.result(LayerAttributes.delete(layerId.name, layerId.zoom), Duration.Inf)
+  def delete(layerId: LayerId): Unit = {
+    logger.debug(s"delete($layerId)")
+    Await.result(LayerAttributes.delete(layerId.name, layerId.zoom), awaitTimeout)
+  }
 
-  def delete(layerId: LayerId, attributeName: String): Unit =
-    Await.result(LayerAttributes.delete(layerId.name, layerId.zoom, attributeName), Duration.Inf)
+  def delete(layerId: LayerId, attributeName: String): Unit = {
+    logger.debug(s"delete($layerId-$attributeName)")
+    Await.result(LayerAttributes.delete(layerId.name, layerId.zoom, attributeName), awaitTimeout)
+  }
 
-  def layerIds: Seq[LayerId] =
-    Await.result(LayerAttributes.layerIds.map(_.map { case (name, zoom) => LayerId(name, zoom) }.toSeq), Duration.Inf)
+  def layerIds: Seq[LayerId] = {
+    logger.debug(s"layerIds($layerIds)")
+    Await.result(LayerAttributes.layerIds.map(_.map { case (name, zoom) => LayerId(name, zoom) }.toSeq), awaitTimeout)
+  }
 
-  def layerIds(layerNames: Set[String]): Seq[LayerId] =
-    Await.result(LayerAttributes.layerIds(layerNames).map(_.map { case (name, zoom) => LayerId(name, zoom) }.toSeq), Duration.Inf)
+  def layerIds(layerNames: Set[String]): Seq[LayerId] = {
+    logger.debug(s"layerIdsWithLayerNames($layerNames)")
+    Await.result(LayerAttributes.layerIds(layerNames).map(
+      _.map { case (name, zoom) => LayerId(name, zoom) }.toSeq), awaitTimeout)
+  }
 
-  def maxZoomsForLayers(layerNames: Set[String]): Map[String, Int] =
-    Await.result(LayerAttributes.maxZoomsForLayers(layerNames), Duration.Inf).toMap
+  def maxZoomsForLayers(layerNames: Set[String]): Future[Option[Map[String, Int]]] = {
```

Comment:
Hmm, that's a good point. Right now everywhere this is called we assume it's blocking I think. Not sure there is a good way around this.